Userland/Libraries/LibWeb/HTML/HTMLElement.cpp

```diff
@@ -608,6 +611,40 @@ TokenizedFeature::NoOpener HTMLElement::get_an_elements_noopener(StringView targ
     return TokenizedFeature::NoOpener::No;
 }
 
+WebIDL::ExceptionOr<JS::NonnullGCPtr<ElementInternals>> HTMLElement::attach_internals()
+{
+    // 1. If this's is value is not null, then throw a "NotSupportedError" DOMException.
+    if (is_value().has_value())
+        return WebIDL::NotSupportedError::create(realm(), "ElementInternals cannot be attached to a customized build-in element"_fly_string);
+
+    // 2. Let definition be the result of looking up a custom element definition given this's node document, its namespace, its local name, and null as the is value.
+    auto definition = document().lookup_custom_element_definition(namespace_uri(), local_name(), is_value());
+
+    // 3. If definition is null, then throw an "NotSupportedError" DOMException.
+    if (!definition)
+        return WebIDL::NotSupportedError::create(realm(), "ElementInternals cannot be attached to an element that is not a custom element"_fly_string);
+
+    // 4. If definition's disable internals is true, then throw a "NotSupportedError" DOMException.
+    if (definition->disable_internals())
+        return WebIDL::NotSupportedError::create(realm(), "ElementInternals are disabled for this custom element"_fly_string);
+
+    // 5. If this's attached internals is non-null, then throw an "NotSupportedError" DOMException.
+    if (m_attached_internals)
+        return WebIDL::NotSupportedError::create(realm(), "ElementInternals already attached"_fly_string);
+
+    // 6. If this's custom element state is not "precustomized" or "custom", then throw a "NotSupportedError" DOMException.
+    if (custom_element_state() != DOM::CustomElementState::Precustomized && custom_element_state() != DOM::CustomElementState::Custom)
```

Comment:
probably not the most interesting optimization for this specific check, but if we add an `AK_ENUM_BITWISE_OPERATORS` for CustomElementState, (and space out its values on powers of two) we could do 

```
if (has_any_flag(custom_element_state(), DOM::CustomElementState::Precustomized | DOM::CustomElementState::Custom))
```

Reply:
> probably not the most interesting optimization for this specific check, but if we add an `AK_ENUM_BITWISE_OPERATORS` for CustomElementState, (and space out its values on powers of two)

Does using `AK_ENUM_BITWISE_OPERATORS` make sense here if `CustomElementState` is only supposed to represent one value at a time? I think something like this would do the same thing:
```c++
if (!first_is_one_of(custom_element_state(), DOM::CustomElementState::Precustomized, DOM::CustomElementState::Custom))
```

Reply:
I changed it to the first_is_one_of code